drivers/sensors/apds9922.c

```diff
@@ -0,0 +1,2598 @@
+        }
+    }
+  else if (fds->priv)
+    {
+      /* This is a request to tear down the poll. */
+
+      struct pollfd **slot = (struct pollfd **)fds->priv;
+      DEBUGASSERT(slot != NULL);
+
+      /* Remove all memory of the poll setup */
+
+      *slot = NULL;
+      fds->priv = NULL;
+    }
+
+out:
+  nxmutex_unlock(&priv->devlock);
+  return ret;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: apds9922_register
+ *
+ * Description:
+ *   Register the APDS9922 character devices.
+ *
+ * Input Parameters:
+ *   devpath_als - The full path to the driver to register for the als,
+ *   e.g., "/dev/als0". If NULL the dreiver will not be registered.
+ *
+ *   devpath_ps - The full path to the driver to register for the als,
+ *   e.g., "/dev/ps0". If NULL the dreiver will not be registered.
+ *
+ *   config - Pointer to the device configuration
+ *
+ * Returned Value:
+ *   Zero (OK) on success; a negated errno value on failure.
+ *
+ ****************************************************************************/
+
+int apds9922_register(FAR const char *devpath_als,
+                      FAR const char *devpath_ps,
+                      FAR struct apds9922_config_s *config)
+{
+  int ret;
+  uint8_t regval;
+
+  /* Initialize the APDS9922 device structure */
+
+  FAR struct apds9922_dev_s *priv;
+
+  priv = kmm_zalloc(sizeof(*priv));
+
+  if (priv == NULL)
+    {
+      snerr("ERROR: Failed to allocate instance\n");
+      return -ENOMEM;
+    }
+
+  nxmutex_init(&priv->devlock);
+
+  priv->config = config;
+
+  /* Reset the device to make it sane */
+
+  apds9922_reset(priv);
+
+  /* Probe APDS9922 device */
+
+  ret = apds9922_probe(priv);
+  if (ret < 0)
+    {
+      goto err_out;
+    }
+
+  /* Enable ALS and/or PS */
+
+  regval = (devpath_ps != NULL)   ? PS_ACTIVE  : 0;
+  regval |= (devpath_als != NULL) ? ALS_ACTIVE : regval;
+  ret = apds9922_i2c_write8(priv, APDS9922_MAIN_CTRL, regval);
+  if (ret < 0)
+    {
+      snerr("ERROR: Failed to enable als and/or ps.\n");
+      goto err_out;
+    }
+
+  /* device interrupts are enabled by default. Disable them. */
+
+  ret = apds9922_als_int_mode(priv, ALS_INT_MODE_DISABLED);
+  if (ret < 0)
+    {
+      snerr("ERROR: Failed to disable ALS interrupts.\n");
+      goto err_out;
+    }
+
+  apds9922_ps_int_mode(priv, PS_INT_MODE_DISABLED);
+  if (ret < 0)
+    {
+      snerr("ERROR: Failed to disable PS interrupts.\n");
+      goto err_out;
+    }
+
+  /* Register the character driver */
+
+  if (devpath_als != NULL)
+    {
+      ret = register_driver(devpath_als, &g_apds9922_alsfops, 0666, priv);
+      if (ret < 0)
+        {
+          snerr("ERROR: Failed to register driver %s: %d\n",
+                 devpath_als, ret);
+          kmm_free(priv);
```

Comment:
Maybe kmm_free should be a part of err_out path as it is missing in many places?

Reply:
Good spot - I'll check,

Reply:
Done